extensions/cornerstone-dicom-sr/src/getSopClassHandlerModule.ts

```diff
@@ -205,115 +207,153 @@ async function _load(displaySet, servicesManager: AppTypes.ServicesManager, exte
+ * @param srDisplaySet - The source display set containing measurements.
+ * @param newDisplaySet - The new display set to check if measurements can be added.
+ * @param dataSource - The data source used to retrieve image IDs.
+ * @param servicesManager - The services manager.
+ */
 function _checkIfCanAddMeasurementsToDisplaySet(
   srDisplaySet,
   newDisplaySet,
   dataSource,
   servicesManager: AppTypes.ServicesManager
 ) {
   const { customizationService } = servicesManager.services;
+
   let unloadedMeasurements = srDisplaySet.measurements.filter(
     measurement => measurement.loaded === false
   );
 
-  if (unloadedMeasurements.length === 0) {
-    // All already loaded!
+  if (
+    unloadedMeasurements.length === 0 ||
+    !(newDisplaySet instanceof ImageSet) ||
+    newDisplaySet.unsupported
+  ) {
     return;
   }
 
-  if ((!newDisplaySet) instanceof ImageSet) {
-    // This also filters out _this_ displaySet, as it is not an ImageSet.
-    return;
-  }
+  const { sopClassUids } = newDisplaySet;
+  // Create a Set for faster lookups
+  const sopClassUidSet = new Set(sopClassUids);
 
-  if (newDisplaySet.unsupported) {
-    return;
-  }
+  // Create a Map to efficiently look up ImageIds by SOPInstanceUID and frame number
+  const imageIdMap = new Map<string, string>();
+  const imageIds = dataSource.getImageIdsForDisplaySet(newDisplaySet);
 
-  const { sopClassUids } = newDisplaySet;
+  for (const imageId of imageIds) {
+    const { SOPInstanceUID, frameNumber } = metadataProvider.getUIDsFromImageID(imageId);
+    const key = `${SOPInstanceUID}:${frameNumber || 1}`;
+    imageIdMap.set(key, imageId);
+  }
 
-  // Check if any have the newDisplaySet is the correct SOPClass.
+  // Filter unloaded measurements based on SOPClassUID
   unloadedMeasurements = unloadedMeasurements.filter(measurement =>
-    measurement.coords.some(coord =>
-      sopClassUids.includes(coord.ReferencedSOPSequence.ReferencedSOPClassUID)
-    )
+    measurement.coords.some(coord => {
+      if (coord.ReferencedSOPSequence === undefined && coord.ReferencedFrameOfReferenceSequence) {
+        for (const [key, imageId] of imageIdMap) {
+          const imageMetadata = metadataProvider.get('instance', imageId);
+          if (imageMetadata.FrameOfReferenceUID !== coord.ReferencedFrameOfReferenceSequence) {
+            continue;
+          }
+
+          const sliceNormal = [0, 0, 0];
+          const orientation = imageMetadata.ImageOrientationPatient;
+          sliceNormal[0] = orientation[1] * orientation[5] - orientation[2] * orientation[4];
+          sliceNormal[1] = orientation[2] * orientation[3] - orientation[0] * orientation[5];
+          sliceNormal[2] = orientation[0] * orientation[4] - orientation[1] * orientation[3];
+
+          let distanceAlongNormal = 0;
+          for (let j = 0; j < 3; ++j) {
+            distanceAlongNormal += sliceNormal[j] * imageMetadata.ImagePositionPatient[j];
+          }
+
+          /** Assuming 2 mm tolerance */
```

Comment:
where is this assumption coming from? can we introduce a epsilon on top?

Reply:
Great catch!